shared/actions/chat.js

```diff
@@ -1044,27 +1031,62 @@ function _unboxedToMessage (message: MessageUnboxed, idx: number, yourName, your
             outboxID,
             key: common.messageID,
           }
-        case CommonMessageType.attachment:
-          // $FlowIssue
-          const preview: Asset = payload.messageBody.attachment.preview
+        case CommonMessageType.attachment: {
+          if (!payload.messageBody.attachment) {
+            throw new Error('empty attachment body')
+          }
+          const attachment: ChatTypes.MessageAttachment = payload.messageBody.attachment
+          const previews = attachment && attachment.previews
+          const preview = previews && previews[0]
           const mimeType = preview && preview.mimeType
           const previewSize = preview && preview.metadata && Constants.parseMetadataPreviewSize(preview.metadata)
 
+          let messageState
+          if (attachment.uploaded) {
+            messageState = 'sent'
+          } else {
+            messageState = common.author === common.you ? 'uploading' : 'placeholder'
+          }
+
           return {
             type: 'Attachment',
             ...common,
-            // $FlowIssue todo fix
-            filename: payload.messageBody.attachment.object.filename,
-            // $FlowIssue todo fix
-            title: payload.messageBody.attachment.object.title,
-            messageState: 'sent',
+            filename: attachment.object.filename,
+            title: attachment.object.title,
+            messageState,
             previewType: mimeType && mimeType.indexOf('image') === 0 ? 'Image' : 'Other',
             previewPath: null,
             hdPreviewPath: null,
             previewSize,
             downloadedPath: null,
             key: common.messageID,
           }
+        }
+        case CommonMessageType.attachmentuploaded: {
+          if (!payload.messageBody.attachmentuploaded) {
+            throw new Error('empty attachmentuploaded body')
+          }
+          const attachmentUploaded: ChatTypes.MessageAttachmentUploaded = payload.messageBody.attachmentuploaded
+          const previews = attachmentUploaded && attachmentUploaded.previews
+          const preview = previews && previews[0]
+          const mimeType = preview && preview.mimeType
+          const previewSize = preview && preview.metadata && Constants.parseMetadataPreviewSize(preview.metadata)
+
+          return {
+            key: common.messageID,
+            messageID: common.messageID,
+            targetMessageID: attachmentUploaded.messageID,
+            timestamp: common.timestamp,
+            type: 'UpdateAttachment',
+            updates: {
+              filename: attachmentUploaded.object.filename,
+              messageState: 'sent',
+              previewType: mimeType && mimeType.indexOf('image') === 0 ? 'Image' : 'Other',
```

Comment:
@MarcoPolo Would it be `image/webp`? Good thought -- should we ticket refinements to the mimetype test?